src/iceberg/table.h

```diff
@@ -35,77 +36,88 @@ class ICEBERG_EXPORT Table {
-  virtual const std::unordered_map<int32_t, std::shared_ptr<SortOrder>>& sort_orders()
-      const = 0;
+  const std::shared_ptr<std::unordered_map<int32_t, std::shared_ptr<SortOrder>>>&
+  sort_orders() const;
 
   /// \brief Return a map of string properties for this table
-  virtual const std::unordered_map<std::string, std::string>& properties() const = 0;
+  const std::unordered_map<std::string, std::string>& properties() const;
 
   /// \brief Return the table's base location
-  virtual const std::string& location() const = 0;
+  const std::string& location() const;
 
-  /// \brief Return the table's current snapshot
-  virtual const std::shared_ptr<Snapshot>& current_snapshot() const = 0;
+  /// \brief Return the table's current snapshot, return NotFoundError if not found
+  Result<std::shared_ptr<Snapshot>> current_snapshot() const;
 
-  /// \brief Get the snapshot of this table with the given id, or null if there is no
-  /// matching snapshot
+  /// \brief Get the snapshot of this table with the given id
   ///
   /// \param snapshot_id the ID of the snapshot to get
-  /// \return the Snapshot with the given id
-  virtual Result<std::shared_ptr<Snapshot>> snapshot(int64_t snapshot_id) const = 0;
+  /// \return the Snapshot with the given id, return NotFoundError if not found
+  Result<std::shared_ptr<Snapshot>> SnapshotById(int64_t snapshot_id) const;
 
   /// \brief Get the snapshots of this table
-  virtual const std::vector<std::shared_ptr<Snapshot>>& snapshots() const = 0;
+  const std::vector<std::shared_ptr<Snapshot>>& snapshots() const;
 
   /// \brief Get the snapshot history of this table
   ///
   /// \return a vector of history entries
-  virtual const std::vector<std::shared_ptr<HistoryEntry>>& history() const = 0;
-
-  /// \brief Create a new table scan for this table
-  ///
-  /// Once a table scan is created, it can be refined to project columns and filter data.
-  virtual std::unique_ptr<TableScan> NewScan() const = 0;
-
-  /// \brief Create a new append API to add files to this table and commit
-  virtual std::shared_ptr<AppendFiles> NewAppend() = 0;
-
-  /// \brief Create a new transaction API to commit multiple table operations at once
-  virtual std::unique_ptr<Transaction> NewTransaction() = 0;
-
-  /// TODO(wgtmac): design of FileIO is not finalized yet. We intend to use an
-  /// IO-less design in the core library.
-  // /// \brief Returns a FileIO to read and write table data and metadata files
-  // virtual std::shared_ptr<FileIO> io() const = 0;
-
-  /// \brief Returns a LocationProvider to provide locations for new data files
-  virtual std::unique_ptr<LocationProvider> location_provider() const = 0;
+  const std::vector<SnapshotLogEntry>& history() const;
+
+  /// \brief Returns a FileIO to read and write table data and metadata files
+  const std::shared_ptr<FileIO>& io() const;
+
+ private:
+  const TableIdentifier identifier_;
+  const std::shared_ptr<TableMetadata> metadata_;
```

Comment:
```suggestion
  std::shared_ptr<TableMetadata> metadata_;
```

I think this will eventually be non-const when `Table::Refresh()` has been added.